usersync/syncersbuilder.go

```diff
@@ -98,28 +101,88 @@ func chooseSyncerConfig(biddersSyncerConfig []namedSyncerConfig) (namedSyncerCon
 		return biddersSyncerConfig[0], nil
 	}
 
-	var bidderNames []string
-	var bidderNamesWithEndpoints []string
-	var syncerConfig namedSyncerConfig
+	var (
+		bidderNames         []string
+		nonAliasBidderNames []string
+		aliasBidderNames    []string
+		syncerConfig        namedSyncerConfig
+		parentBidderNameMap = make(map[string]struct{})
+	)
+
 	for _, bidder := range biddersSyncerConfig {
 		bidderNames = append(bidderNames, bidder.name)
 		if bidder.cfg.IFrame != nil || bidder.cfg.Redirect != nil {
-			bidderNamesWithEndpoints = append(bidderNamesWithEndpoints, bidder.name)
+			if len(bidder.bidderInfo.AliasOf) > 0 {
+				aliasBidderNames = append(aliasBidderNames, bidder.name)
+				parentBidderNameMap[bidder.bidderInfo.AliasOf] = struct{}{}
+			} else {
+				nonAliasBidderNames = append(nonAliasBidderNames, bidder.name)
+			}
 			syncerConfig = bidder
 		}
 	}
 
-	if len(bidderNamesWithEndpoints) == 0 {
+	// check if bidders have same syncer key but no endpoints
+	if len(nonAliasBidderNames)+len(aliasBidderNames) == 0 {
 		sort.Strings(bidderNames)
 		bidders := strings.Join(bidderNames, ", ")
 		return namedSyncerConfig{}, fmt.Errorf("bidders %s share the same syncer key, but none define endpoints (iframe and/or redirect)", bidders)
 	}
 
-	if len(bidderNamesWithEndpoints) > 1 {
-		sort.Strings(bidderNamesWithEndpoints)
-		bidders := strings.Join(bidderNamesWithEndpoints, ", ")
+	// check if non-alias bidders have same syncer key
+	if len(nonAliasBidderNames) > 1 {
+		sort.Strings(nonAliasBidderNames)
+		bidders := strings.Join(nonAliasBidderNames, ", ")
 		return namedSyncerConfig{}, fmt.Errorf("bidders %s define endpoints (iframe and/or redirect) for the same syncer key, but only one bidder is permitted to define endpoints", bidders)
 	}
 
+	if len(parentBidderNameMap) != 0 {
+		// check if alias bidders of different parent have same syncer key
+		if len(parentBidderNameMap) > 1 {
+			sort.Strings(aliasBidderNames)
+			return namedSyncerConfig{}, fmt.Errorf("either alias bidders %s defines endpoints (iframe and/or redirect) for the same syncer key used by multiple bidders, but only one bidder is permitted to define endpoints", strings.Join(aliasBidderNames, ", "))
+		}
+
+		// check if aliases of same parent and non-alias bidder have same syncer key
+		if _, ok := parentBidderNameMap[nonAliasBidderNames[0]]; !ok {
+			sort.Strings(aliasBidderNames)
+			return namedSyncerConfig{}, fmt.Errorf("either alias bidders %s defines endpoints (iframe and/or redirect) for the same syncer key used by other bidder, but only one bidder is permitted to define endpoints", strings.Join(aliasBidderNames, ", "))
+		}
```

Comment:
This is getting a bit confusing to follow. This logic could be left alone if you resolve the alias user sync keys up front:

- If the alias defines it's own syncer or syncer key, do nothing. The validation in this function will ensure the syncer rules are adhered.
- if the alias does not define it's own syncer or syncer key but it's parent does, set the syncer key to match the parent. This is either the parent's syncer key or the parent's bidder name (the default implicit syncer key) - which you already have coded.

This may simplify the validation logic.

Reply:
@onkarvhanumante was this comment addressed by you? Were you able to see if this validation logic can be simplified?

Reply:
> @onkarvhanumante was this comment addressed by you? Were you able to see if this validation logic can be simplified?

@AlexBVolcy,  had some discussion with @SyntaxNode and  have evaluated few cases to understand what syncer key should be assigned to alias and parent bidder. Will push updated changes and inform here

Reply:
@AlexBVolcy  @gargcreation1992  PR is updated with latest commits. All previous commits are reverted and https://github.com/prebid/prebid-server/pull/3082/commits/dd12c58edf74bf6ce2a1e2d4e493fe2115138398 is the one to review.

Refer https://github.com/prebid/prebid-server/pull/3082#issue-1885175603 for testing details